### PR TITLE
[ Coverity ] Fix Race Condition issue about databuffer

### PR DESCRIPTION
**Pull request by jijoongmoon**
. Add Lock to set data ready flag.
. change order to set the condition

. Fix undeterministic behavior of databuffer
. related issues : #427, #464

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/src/databuffer.cpp [hunk 1]
```diff
@@ -187,12 +185,15 @@ bool DataBuffer::getDataFromBuffer(BufferType type, vec_4d &outVec,
...
     std::vector<int> list;
-    std::unique_lock<std::mutex> ultrain(readyTrainData);
-    cv_train.wait(ultrain, [this]() -> int { return trainReadyFlag; });
-
-    if (train_data.size() < batch_size || trainReadyFlag == DATA_ERROR ||
-        trainReadyFlag == DATA_END) {
-      return false;
+    while (true) {
+      std::unique_lock<std::mutex> ultrain(readyTrainData);
+      cv_train.wait(ultrain, [this]() -> int { return trainReadyFlag; });
```
Comment (zhoonit) [inline nntrainer/src/databuffer.cpp:190]: Could you check if this is valid?
https://en.cppreference.com/w/cpp/thread/condition_variable/wait

```suggestion
      cv_train.wait(ultrain, [this]() -> bool{ return trainReadyFlag; });
```

File: nntrainer/src/databuffer.cpp [hunk 2]
```diff
@@ -187,12 +185,15 @@ bool DataBuffer::getDataFromBuffer(BufferType type, vec_4d &outVec,
...
+      if (trainReadyFlag == DATA_ERROR || trainReadyFlag == DATA_END) {
```
Comment (zhoonit) [inline nntrainer/src/databuffer.cpp:191]:
```suggestion
      if (trainReadyFlag == DATA_ERROR || trainReadyFlag == DATA_END || train_data.empy()) {
```

to eliminate edge cases? (and for other cases)

File: nntrainer/src/databuffer.cpp [hunk 3]
```diff
@@ -374,10 +383,15 @@ int DataBuffer::init() {
   this->cur_val_bufsize = 0;
   this->cur_test_bufsize = 0;
 
+  readyTrainData.lock();
```
Comment (kparichay) [inline nntrainer/src/databuffer.cpp:386]: do we need locking here? this is init and no thread is active yet.

Comment (jijoongmoon) [reply nntrainer/src/databuffer.cpp:386]: coverity says we need.

File: nntrainer/src/databuffer.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/src/databuffer.cpp:191]: it will be checked at line 194

File: nntrainer/src/databuffer.cpp (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/src/databuffer.cpp:190]: fixed. thanks.

File: nntrainer/src/databuffer.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/src/databuffer.cpp:386]: ok

**Review by kparichay — APPROVED** [review]
LGTM

**Review by zhoonit — APPROVED** [review]
LGTM

**Pull request merged**